src/bin/cargo-miri.rs

```diff
@@ -34,6 +48,149 @@ fn show_version() {
+                    "could not read current directory",
+                );
+                let package_manifest_directory = package_manifest_path.parent().expect(
+                    "could not find parent directory of package manifest",
+                );
+                package_manifest_directory == current_dir
+            }
+        })
+        .expect("could not find matching package");
+    let package = metadata.packages.remove(package_index);
+
+    // Finally we got the list of targets to build
+    package.targets.into_iter()
+}
+
+fn ask(question: &str) {
+    let mut buf = String::new();
+    print!("{} [Y/n] ", question);
+    io::stdout().flush().unwrap();
+    io::stdin().read_line(&mut buf).unwrap();
+    let answer = match buf.trim().to_lowercase().as_ref() {
+        "" | "y" | "yes" => true,
+        "n" | "no" => false,
+        a => show_error(format!("I do not understand `{}`", a))
+    };
+    if !answer {
+        show_error(format!("Aborting as per your request"))
+    }
+}
+
+/// Perform the setup requires to make `cargo miri` work: Getting a custom-built libstd. Then sets MIRI_SYSROOT.
+/// Skipped if MIRI_SYSROOT is already set, in that case we expect the user has done all this already.
+fn setup(ask_user: bool) {
+    if std::env::var("MIRI_SYSROOT").is_ok() {
+        return;
+    }
+
+    // First, we need xargo
+    if Command::new("xargo").arg("--version").output().is_err()
+    {
+        if ask_user {
+            ask("It seems you do not have xargo installed. I will run `cargo install xargo`. Proceed?");
+        } else {
+            println!("Installing xargo: `cargo install xargo`");
+        }
+        if !Command::new("cargo").args(&["install", "xargo"]).status().unwrap().success() {
+            show_error(format!("Failed to install xargo"));
+        }
+    }
+
+    // Then, we also need rust-src.  Let's see if it is already installed.
+    let sysroot = Command::new("rustc").args(&["--print", "sysroot"]).output().unwrap().stdout;
+    let sysroot = std::str::from_utf8(&sysroot[..]).unwrap();
+    let src = Path::new(sysroot.trim_end_matches('\n')).join("lib").join("rustlib").join("src");
+    if !src.exists() {
+        if ask_user {
+            ask("It seems you do not have the rust-src component installed. I will run `rustup component add rust-src`. Proceed?");
+        } else {
+            println!("Installing rust-src component: `rustup component add rust-src`");
+        }
+        if !Command::new("rustup").args(&["component", "add", "rust-src"]).status().unwrap().success() {
+            show_error(format!("Failed to install rust-src component"));
+        }
+    }
+
+    // Next, we need our own libstd. We will do this work in ~/.miri.
+    let dir = dirs::home_dir().unwrap().join(".miri");
+    if !dir.exists() {
+        fs::create_dir(&dir).unwrap();
+    }
+    // The interesting bit: Xargo.toml
+    File::create(dir.join("Xargo.toml")).unwrap()
+        .write_all(br#"
```

Comment:
maybe put these into files in a separate directory and use `include_bytes!`?

Reply:
You think that's worth it? They are rather short.